presto-common/src/main/java/com/facebook/presto/common/block/AbstractArrayBlock.java

```diff
@@ -51,35 +55,55 @@ public String getEncodingName()
     }
 
     @Override
-    public Block copyPositions(int[] positions, int offset, int length)
+    public final Block copyPositions(int[] positions, int offset, int length)
     {
         checkArrayRange(positions, offset, length);
 
+        int oldOffsetBase = getOffsetBase();
+        int[] oldOffsets = getOffsets();
         int[] newOffsets = new int[length + 1];
-        boolean[] newValueIsNull = new boolean[length];
-
-        IntArrayList valuesPositions = new IntArrayList();
-        int newPosition = 0;
-        for (int i = offset; i < offset + length; ++i) {
-            int position = positions[i];
-            if (isNull(position)) {
-                newValueIsNull[newPosition] = true;
-                newOffsets[newPosition + 1] = newOffsets[newPosition];
+        boolean[] newValueIsNull = mayHaveNull() ? new boolean[length] : null;
+        int currentOffset = 0;
```

Comment:
With my following suggested rewrites, this line will be moved to after the first for loop